src/rascal/math/kvec_generator.hh

```diff
@@ -0,0 +1,229 @@
+       * @param need_origin false by default.
+       *     If set to true, the list of found k-vectors will also
+       *     contain the origin k=(0,0,0), which will then be stored
+       *     as the very first element in the matrix of found vectors.
+       */
+      explicit Kvectors(double cutoff, Eigen::Matrix3d basisvectors,
+                        bool is_reciprocal_cell = false,
+                        bool need_origin = false) {
+        // Set internal variables apart from cell
+        this->kcutoff = cutoff;
+        this->include_origin = need_origin;
+
+        /* Store provided basis vectors into the internal variable basisvecs.
+         * We need to distinguish whether the provided vectors are already
+         * reciprocal cell vectors or need to be converted first
+         */
+        if (not is_reciprocal_cell) {  // real space cell is given
+          // Create reciprocal cell from real space cell
+          Eigen::Matrix3d tcell = basisvectors.transpose();
+          this->basisvecs = 2.0 * M_PI * tcell.inverse();
+        } else {  // cell of reciprocal space is already given
+          this->basisvecs = basisvectors;
+        }
+
+        // Run the precompute function described below
+        this->precompute();
+      }  // end constructor
+
+      /**
+       * Using the provided parameters in the constructor, find
+       * all the k-vectors within the cutoff radius and store them
+       * as well as derived quantities that can be returned to the user.
+       */
+      void precompute();
+
+      // FUNCTIONS FOR USER
+
+      /** Returns number of vectors found within cutoff radius
+       * without double counting pairs related by inversion.
+       * This is the actual number that can be used to iterate over all
+       * vectors and its norms obtained from get_kvectors and get_norms.
+       */
+      size_t get_numvectors() const { return this->numvectors; }
+
+      /** Returns number of vectors found within cutoff radius
+       * counting pairs related by inversion separately.
+       * If the origin is excluded, the number without origin is doubled
+       * and returned. Otherwise, the origin is counted once.
+       */
+      size_t get_numvectors_all() const {
+        if (this->include_origin) {
+          return 2 * this->numvectors - 1;
+        } else {
+          return 2 * this->numvectors;
+        }
+      }
+
+      /** Returns the Eigen matrix of shape this->numvectors x 3 containing
+       * all k-vectors within cutoff radius without double counting,
+       * where row(i) = i-th vector. If the origin k=(0,0,0) is included,
+       * it is always stored in row(0).
+       */
+      Matrix_t get_kvectors() const { return this->kvecs; }
+
+      /** Returns the Eigen vector of length this->numvectors containing
+       * the norm of all k-vectors within the cutoff without double counting.
+       */
+      Vector_t get_norms() const { return this->kvecnorms; }
```

Comment:
I'm not familiar with Eigen enough, but this code looks like it is making a copy. Could we write these two functions to return const references instead?

Reply:
As far as I remember one can do this as in
https://github.com/cosmo-epfl/librascal/blob/master/src/rascal/representations/calculator_spherical_expansion.hh#L1368
https://github.com/cosmo-epfl/librascal/blob/master/src/rascal/representations/calculator_spherical_expansion.hh#L1374

using definitions
https://github.com/cosmo-epfl/librascal/blob/master/src/rascal/representations/calculator_spherical_expansion.hh#L222